src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+                        .into()
+                    ),
+                    (
+                        "d",
+                        FSEntry::Dir(DirEntry {
+                            properties: Some(Mode(4, true)),
+                            children: BTreeMap::new(),
+                        })
+                        .into()
+                    ),
+                    ("e", FSEntry::File(Mode(5, false)).into())
+                ]
+                .into_iter()
+                .collect()
+            );
+        }
+    }
+}
+
+pub mod handle_creation {
+    use displaydoc::Display;
+    use thiserror::Error;
+
+    use std::cmp;
+    use std::collections::{HashMap, VecDeque};
+    use std::fs;
+    use std::hash;
+    use std::io;
+    use std::path::{Path, PathBuf};
+
+    use crate::types::ZipFileData;
+
+    use super::path_splitting::{DirEntry, FSEntry};
+
+    /// Errors encountered when creating output handles for extracting entries to.
+    #[derive(Debug, Display, Error)]
+    pub enum HandleCreationError {
+        /// i/o error: {0}
+        Io(#[from] io::Error),
+    }
+
+    /// Wrapper for memory location of the ZipFileData in Shared.
+    ///
+    /// Enables quick comparison and hash table lookup without needing to implement Hash/Eq for
+    /// ZipFileData more generally.
+    #[derive(Debug)]
+    pub(crate) struct ZipDataHandle<'a>(&'a ZipFileData);
+
+    impl<'a> ZipDataHandle<'a> {
+        #[inline(always)]
+        const fn ptr(&self) -> *const ZipFileData {
+            self.0
+        }
+
+        #[inline(always)]
+        pub const fn wrap(data: &'a ZipFileData) -> Self {
+            Self(data)
+        }
+    }
+
+    impl<'a> cmp::PartialEq for ZipDataHandle<'a> {
+        #[inline(always)]
+        fn eq(&self, other: &Self) -> bool {
+            self.ptr() == other.ptr()
+        }
+    }
+
+    impl<'a> cmp::Eq for ZipDataHandle<'a> {}
+
+    impl<'a> hash::Hash for ZipDataHandle<'a> {
+        #[inline(always)]
+        fn hash<H: hash::Hasher>(&self, state: &mut H) {
+            self.ptr().hash(state);
+        }
+    }
+
+    /* TODO: figure out how to handle symlinks! These are especially difficult because:
+     * (1) windows symlinks files and directories differently, and only on newer windows versions,
+     * (2) later entries in the zip may refer to symlink paths from earlier in the zip.
+     *
+     * Of these issues, (2) is more difficult and intrinsic to the problem space. In order to
+     * correctly extract symlinks in a pipelined/parallel fashion, we need to identify these order
+     * dependencies and schedule the symlink dereference (reading the target value from the zip)
+     * before we create any directories or allocate any output file handles that dereference that
+     * symlink. This is less of a problem with the synchronous in-order extraction because it
+     * creates any symlinks immediately (it imposes a total ordering dependency over all entries).
+     */
```

Comment:
Use [futures::future::Shared](https://docs.rs/futures/latest/futures/future/struct.Shared.html) to handle this; it's what I used for a dependency DAG of parallel tasks in https://github.com/pr0methean/OcHd-RustBuild. It looks like you're reinventing the wheel here, so I won't review the rest of this module for now.

Reply:
Hm, ok. I was hoping to leave symlink support for a follow-up PR with this TODO specifically to avoid having to construct the dependency DAG of tasks you describe. Would you accept that compromise?

Reply:
I have added a note here linking to the shared future object!